lib/web/ui/join_token.go

```diff
@@ -0,0 +1,73 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ui
+
+import (
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+type JoinToken struct {
+	// Id is the name of the token
+	Id string `json:"id"`
+	// SafeName returns the name of the token, sanitized appropriately for
+	// join methods where the name is secret.
+	SafeName string `json:"safeName"`
+	// Expiry is the time that the token resource expires
+	Expiry time.Time `json:"expiry"`
+	// Roles are the roles granted to the token
+	Roles types.SystemRoles `json:"roles"`
+	// IsStatic is true if the token is statically configured
+	IsStatic bool `json:"isStatic"`
+	// Method is the join method that the token supports
+	Method types.JoinMethod `json:"method"`
+	// SuggestedLabels contains the set of labels we expect the node to set when using this token
+	SuggestedLabels []Label `json:"suggestedLabels,omitempty"`
+	// AllowRules is a list of allow rules
+	AllowRules []string `json:"allowRules,omitempty"`
+}
+
+func MakeJoinToken(token types.ProvisionToken) JoinToken {
+	labels := token.GetSuggestedLabels()
+	suggestedLabels := make([]Label, len(labels))
```

Comment:
```suggestion
	suggestedLabels := make([]Label, 0, len(labels))
```

Otherwise `suggestedLabels` is going to be twice as long as you expect.